Sources/core/objects/Human.cpp

```diff
@@ -82,6 +84,274 @@ bool Human::TryMove(Dir direct)
         {
             PlaySoundIfVisible("glass_step.ogg", GetOwner().Id());
         }
+        if (pulled_object_)
+        {
+            if (!CanTouch(pulled_object_))
```

Comment:
Overall, it seems overcomplicated.
1. It is not needed to call `GetRoot()->GetX()` because just `GetX()` should return same value.
2. It is not needed manually check passable levels, remove and add the pulled object. `TryMove` function do that by itself.
3. It is better to work with previous position of the human. You should store human position before the move (before `IMob::TryMove`), and then if the human movement was successfull, you should move the pulled object to the previous position of the human.
4. Also `CanTouch` also should be before the human movement.